### PR TITLE
WIP Refactoring knockoff part + improve docstring

### PR DESCRIPTION
**Pull request by tbng**
Todo:
- [ ] Knockoff sphinx example
- [ ] improve docstring

### CODE REVIEW

File: hidimstat/knockoffs.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    Parameters
+    ----------
+    X : 2D ndarray (n_samples, n_features)
+        design matrix
+
+    y : 1D ndarray (n_samples, )
+        response vector
+
+    fdr : float, optional
+        desired controlled FDR level
```
Comment (bthirion) [inline hidimstat/knockoffs.py:38]: range [0, 1]

File: hidimstat/knockoffs.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+    Returns
+    -------
+    selected : 1D array, int
+        vector of index of selected variables
+
+    test_score : 1D array, (n_features, )
+        vector of test statistic
+
+    thres : float
```
Comment (bthirion) [inline hidimstat/knockoffs.py:71]: Call it `threshold`

File: hidimstat/knockoffs.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+    method : str, optional
+        knockoff construction methods, either equi for equi-correlated knockoff
+        or sdp for optimization scheme
+
+    statistics : str, optional
+        method to calculate knockoff test score
+
+    shrink : bool, optional
+        whether to shrink the empirical covariance matrix
```
Comment (bthirion) [inline hidimstat/knockoffs.py:52]: Default is False ?

File: hidimstat/knockoffs.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    X_tilde = gaussian_knockoff_generation(X, mu, Sigma, memory=memory,
+                                           method=method, seed=seed)
+    test_score = memory.cache(
+        stat_coef_diff, ignore=['n_jobs', 'joblib_verbose'])(
+        X, X_tilde, y, method=statistics, n_jobs=n_jobs)
+    thres = _coef_diff_threshold(test_score, fdr=fdr, offset=offset)
+
+    selected = np.where(test_score >= thres)[0]
+
+    if verbose:
```
Comment (bthirion) [inline hidimstat/knockoffs.py:94]: Not a good strategy. May always return the 4 items.

File: hidimstat/knockoffs.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+    statistics : str, optional
+        method to calculate knockoff test score
+
+    adaptive_aggregation : bool, optional
+        whether to use adaptive quantile aggregation scheme for bootstrapping 
+        p-values, for more info see Meinhausen et al. (2009)
+        <https://www.tandfonline.com/doi/abs/10.1198/jasa.2009.tm08647>
+
+    shrink : bool, optional
```
Comment (bthirion) [inline hidimstat/knockoffs.py:143]: clash with cov_estimator argument ?

File: hidimstat/knockoffs.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        whether to shrink the empirical covariance matrix
+
+    centered : bool, optional
+        whether to standardize the data before doing the inference procedure
+
+    cov_estimator : str, optional
```
Comment (bthirion) [inline hidimstat/knockoffs.py:149]: Give the alternatives

**Comment by ja-che** [comment]
@tbng I have moved the KO example in the folder ``examples_not_exhibited``. After finishing this PR you could revise your the KO example in another PR, adding more commentaries and making sure that it takes no more than 5-10 minutes to run.

**Comment by ja-che** [comment]
@tbng Concerning this PR, it would nice to have all the modules in the ``hidimstat`` folder (not in ``hidimstat --> knockoffs``). Also, I believe that some functions are duplicated such as simulations and p-value aggregation. Please, it would be nice to integrate as much as possible the functions that are already in ``hidimstat`` in the knockoffs code. We can discuss that if needed !

**Pull request closed without merging**